src/Bicep.LangServer/Completions/ModuleReferenceCompletionProvider.cs

```diff
@@ -0,0 +1,543 @@
+
+        private static readonly Regex ACRModuleRegistryWithoutAlias = new Regex(@"br:(?<registry>(.*).azurecr.io)/", RegexOptions.Compiled | RegexOptions.ExplicitCapture | RegexOptions.IgnoreCase);
+        private static readonly Regex MCRModuleRegistryWithAlias = new Regex(@"br/public:(?<filePath>(.*?)):", RegexOptions.Compiled | RegexOptions.ExplicitCapture | RegexOptions.IgnoreCase);
+        private static readonly Regex MCRModuleRegistryWithoutAlias = new Regex(@"br:MCR.microsoft.com/bicep/(?<filePath>(.*?)):", RegexOptions.Compiled | RegexOptions.ExplicitCapture | RegexOptions.IgnoreCase);
+        private static readonly Regex ModuleRegistryAliasWithPath = new Regex(@"'br/(.*):(?<filePath>(.*?)):", RegexOptions.Compiled | RegexOptions.ExplicitCapture | RegexOptions.IgnoreCase);
+
+        public ModuleReferenceCompletionProvider(
+            IAzureContainerRegistryNamesProvider azureContainerRegistryNamesProvider,
+            IConfigurationManager configurationManager,
+            IModulesMetadataProvider modulesMetadataProvider,
+            ISettingsProvider settingsProvider,
+            ITelemetryProvider telemetryProvider)
+        {
+            this.azureContainerRegistryNamesProvider = azureContainerRegistryNamesProvider;
+            this.configurationManager = configurationManager;
+            this.modulesMetadataProvider = modulesMetadataProvider;
+            this.settingsProvider = settingsProvider;
+            this.telemetryProvider = telemetryProvider;
+        }
+
+        public async Task<IEnumerable<CompletionItem>> GetFilteredCompletions(Uri templateUri, BicepCompletionContext context)
+        {
+            var replacementText = string.Empty;
+
+            if (context.ReplacementTarget is Token token)
+            {
+                replacementText = token.Text;
+            }
+
+            return GetBicepRegistryAndTemplateSpecShemaCompletions(context, replacementText)
+                .Concat(await GetPathCompletions(context, replacementText, templateUri))
+                .Concat(await GetMCRModuleRegistryVersionCompletions(context, replacementText, templateUri))
+                .Concat(await GetRegistryCompletions(context, replacementText, templateUri));
+        }
+
+        // Handles bicep registry and template spec schema completions.
+        private IEnumerable<CompletionItem> GetBicepRegistryAndTemplateSpecShemaCompletions(BicepCompletionContext context, string replacementText)
+        {
+            if (!context.Kind.HasFlag(BicepCompletionContextKind.ModulePath))
+            {
+                return Enumerable.Empty<CompletionItem>();
+            }
+
+            if (!string.IsNullOrWhiteSpace(replacementText.Trim('\'')))
+            {
+                return Enumerable.Empty<CompletionItem>();
+            }
+
+            List<CompletionItem> completionItems = new List<CompletionItem>();
+            foreach (var kvp in BicepRegistryAndTemplateSpecShemaCompletionLabelsWithDetails)
```

Comment:
>BicepRegistryAndTemplateSpecShemaCompletionLabelsWithDetails [](http://example.com/codeflow?start=32&length=60)

Btw, do we offer `br/` and `ts/` completions if there are no aliases for them in the config file?

Reply:
Good question. I think br/ makes sense, since we would continue it with ```public:```. I made ts/ conditional here:
https://github.com/Azure/bicep/commit/dc29d71dc6f09039b3378b84337b68d08d866fad

Also updated flaky tests as part of this.

Reply:
If there are no aliases configured in bicepconfig.json, there's no valid value you can put after br/ or ts/. So should we even offer those completions in this case? (We can open an issue to track this one if you agree that it's an issue, though.)

Reply:
I think I am missing something here. I made ```ts/``` conditional - show it only if it's configured in bicepconfig.json. 

If there's no alias configured for ```br/```, shouldn't we still offer default ```public```?  I am referring to below scenario where there's no bicepconfig.json

![ShowCompletionForbr](https://user-images.githubusercontent.com/30270536/222524696-ad63a3fb-3d81-4964-9340-1c68626f1271.gif)

Let me know if it's not the case, I'll open issue to track it.

Reply:
Got it. Forgot about public being available by default.